doc/mtree_cksum.3
.\"
.\" Copyright (c) 2015 Michal Ratajsky <michal@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd August 10, 2015
.Dt MTREE_CKSUM 3
.Os
.Sh NAME
.Nm mtree_cksum
.Nd calculate POSIX 1003.2 checksum
.Sh LIBRARY
libmtree
.Sh SYNOPSIS
.In mtree.h
.Ft struct mtree_cksum *
.Fn mtree_cksum_create "uint32_t init"
.Ft void
.Fn mtree_cksum_update "struct mtree_cksum *cksum " "const unsigned char *data" "size_t len"
.Ft uint32_t
.Fn mtree_cksum_get_result "struct mtree_cksum *cksum"
.Ft void
.Fn mtree_cksum_reset "struct mtree_cksum *cksum" "uint32_t init"
.Ft void
.Fn mtree_cksum_free "struct mtree_cksum *cksum"
.In mtree.h
.In mtree_file.h
.Ft int
.Fn mtree_cksum_path "const char *path" "uint32_t *crc"
.Ft int
.Fn mtree_cksum_fd "int fd" "uint32_t *crc"
.Sh DESCRIPTION
These functions provide support for calculating the POSIX 1003.2 checksum. In
mtree spec entries, this number is used as a value of the
.Em checksum
keyword.
.Pp
It is seldom necessary to use these functions directly. Both
.Xr mtree_spec 3
and
.Xr mtree_entry 3
provide facilities for calculating the checksum and including it in the
.Em checksum
keyword of mtree spec entries.
.Pp
Use
.Fn mtree_cksum_create
to create a new
.Tn mtree_cksum
structure. The
.Fa init
argument specifies the initial value of the checksum. Unless some special
initial value is needed, use the constant
.Dv MTREE_CKSUM_DEFAULT_INIT
to zero initialize the checksum.
.Pp
The
.Fn mtree_cksum_update
function should then be called repeatedly to update the checksum with
.Fa data
of size
.Fa len .
.Pp
Use
.Fn mtree_cksum_get_result
to finalize the computation and retrieve the resulting checksum. The finalization
step uses the accumulated length of input to update the checksum at the end of
the computation. Therefore it is an error to call this function to retrieve an
intermediate result before the whole input has been passed to
.Fn mtree_cksum_update .
Once
.Fn mtree_cksum_get_result
is called, no further updates to the checksum are possible and any further uses
of
.Fn mtree_cksum_update
will not affect the checksum.
.Pp
The
.Tn mtree_cksum
structure can be set to its initial state using the
.Fn mtree_cksum_reset
function. The
.Fa init
argument is the initial value of the checksum, just as in
.Fn mtree_cksum_create .
.Pp
Use
.Fn mtree_cksum_free
to release the resources allocated for the structure.
.Pp
The
.Fn mtree_cksum_path
and
.Fn mtree_cksum_fd
functions calculate the checksum by reading input from the given file path,
or an open file descriptor respectively. They both store the resulting checksum
in
.Fa crc .
.Sh RETURN VALUE
The
.Fn mtree_cksum_create
function returns a pointer to a newly allocated
.Tn mtree_cksum
structure, or
.Dv NULL
if memory allocation fails.
.Pp
The
.Fn mtree_cksum_get_result
function returns the calculated checksum.
.Pp
Both
.Fn mtree_cksum_file
and
.Fn mtree_cksum_fd
return zero on success. On error, they return -1 and set
.Va errno
appropriately.
.Sh ERRORS
The
.Fn mtree_cksum_fd
function may fail and set
.Va errno
for any of the errors specified for the routines
.Xr malloc 3
and
.Xr read 2 .
.sp
The
.Fn mtree_cksum_path
function may also fail and set
.Va errno
for any of the errors specified for the routine
.Xr open 2 .
.Sh SEE ALSO
.Xr cksum 1 ,
.Xr mtree_entry 3 ,
.Xr mtree_spec 3
.Sh AUTHORS
.An -nosplit
The
.Nm libmtree
library was written by
.An Michal Ratajsky Aq michal@FreeBSD.org .


libmtree/mtree_cksum.c
/*-
 * Copyright (c) 2015 Michal Ratajsky <michal@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "mtree.h"
#include "mtree_file.h"
#include "mtree_private.h"

static const uint32_t crctab[] = {
	0x0,
	0x04c11db7, 0x09823b6e, 0x0d4326d9, 0x130476dc, 0x17c56b6b,
	0x1a864db2, 0x1e475005, 0x2608edb8, 0x22c9f00f, 0x2f8ad6d6,
	0x2b4bcb61, 0x350c9b64, 0x31cd86d3, 0x3c8ea00a, 0x384fbdbd,
	0x4c11db70, 0x48d0c6c7, 0x4593e01e, 0x4152fda9, 0x5f15adac,
	0x5bd4b01b, 0x569796c2, 0x52568b75, 0x6a1936c8, 0x6ed82b7f,
	0x639b0da6, 0x675a1011, 0x791d4014, 0x7ddc5da3, 0x709f7b7a,
	0x745e66cd, 0x9823b6e0, 0x9ce2ab57, 0x91a18d8e, 0x95609039,
	0x8b27c03c, 0x8fe6dd8b, 0x82a5fb52, 0x8664e6e5, 0xbe2b5b58,
	0xbaea46ef, 0xb7a96036, 0xb3687d81, 0xad2f2d84, 0xa9ee3033,
	0xa4ad16ea, 0xa06c0b5d, 0xd4326d90, 0xd0f37027, 0xddb056fe,
	0xd9714b49, 0xc7361b4c, 0xc3f706fb, 0xceb42022, 0xca753d95,
	0xf23a8028, 0xf6fb9d9f, 0xfbb8bb46, 0xff79a6f1, 0xe13ef6f4,
	0xe5ffeb43, 0xe8bccd9a, 0xec7dd02d, 0x34867077, 0x30476dc0,
	0x3d044b19, 0x39c556ae, 0x278206ab, 0x23431b1c, 0x2e003dc5,
	0x2ac12072, 0x128e9dcf, 0x164f8078, 0x1b0ca6a1, 0x1fcdbb16,
	0x018aeb13, 0x054bf6a4, 0x0808d07d, 0x0cc9cdca, 0x7897ab07,
	0x7c56b6b0, 0x71159069, 0x75d48dde, 0x6b93dddb, 0x6f52c06c,
	0x6211e6b5, 0x66d0fb02, 0x5e9f46bf, 0x5a5e5b08, 0x571d7dd1,
	0x53dc6066, 0x4d9b3063, 0x495a2dd4, 0x44190b0d, 0x40d816ba,
	0xaca5c697, 0xa864db20, 0xa527fdf9, 0xa1e6e04e, 0xbfa1b04b,
	0xbb60adfc, 0xb6238b25, 0xb2e29692, 0x8aad2b2f, 0x8e6c3698,
	0x832f1041, 0x87ee0df6, 0x99a95df3, 0x9d684044, 0x902b669d,
	0x94ea7b2a, 0xe0b41de7, 0xe4750050, 0xe9362689, 0xedf73b3e,
	0xf3b06b3b, 0xf771768c, 0xfa325055, 0xfef34de2, 0xc6bcf05f,
	0xc27dede8, 0xcf3ecb31, 0xcbffd686, 0xd5b88683, 0xd1799b34,
	0xdc3abded, 0xd8fba05a, 0x690ce0ee, 0x6dcdfd59, 0x608edb80,
	0x644fc637, 0x7a089632, 0x7ec98b85, 0x738aad5c, 0x774bb0eb,
	0x4f040d56, 0x4bc510e1, 0x46863638, 0x42472b8f, 0x5c007b8a,
	0x58c1663d, 0x558240e4, 0x51435d53, 0x251d3b9e, 0x21dc2629,
	0x2c9f00f0, 0x285e1d47, 0x36194d42, 0x32d850f5, 0x3f9b762c,
	0x3b5a6b9b, 0x0315d626, 0x07d4cb91, 0x0a97ed48, 0x0e56f0ff,
	0x1011a0fa, 0x14d0bd4d, 0x19939b94, 0x1d528623, 0xf12f560e,
	0xf5ee4bb9, 0xf8ad6d60, 0xfc6c70d7, 0xe22b20d2, 0xe6ea3d65,
	0xeba91bbc, 0xef68060b, 0xd727bbb6, 0xd3e6a601, 0xdea580d8,
	0xda649d6f, 0xc423cd6a, 0xc0e2d0dd, 0xcda1f604, 0xc960ebb3,
	0xbd3e8d7e, 0xb9ff90c9, 0xb4bcb610, 0xb07daba7, 0xae3afba2,
	0xaafbe615, 0xa7b8c0cc, 0xa379dd7b, 0x9b3660c6, 0x9ff77d71,
	0x92b45ba8, 0x9675461f, 0x8832161a, 0x8cf30bad, 0x81b02d74,
	0x857130c3, 0x5d8a9099, 0x594b8d2e, 0x5408abf7, 0x50c9b640,
	0x4e8ee645, 0x4a4ffbf2, 0x470cdd2b, 0x43cdc09c, 0x7b827d21,
	0x7f436096, 0x7200464f, 0x76c15bf8, 0x68860bfd, 0x6c47164a,
	0x61043093, 0x65c52d24, 0x119b4be9, 0x155a565e, 0x18197087,
	0x1cd86d30, 0x029f3d35, 0x065e2082, 0x0b1d065b, 0x0fdc1bec,
	0x3793a651, 0x3352bbe6, 0x3e119d3f, 0x3ad08088, 0x2497d08d,
	0x2056cd3a, 0x2d15ebe3, 0x29d4f654, 0xc5a92679, 0xc1683bce,
	0xcc2b1d17, 0xc8ea00a0, 0xd6ad50a5, 0xd26c4d12, 0xdf2f6bcb,
	0xdbee767c, 0xe3a1cbc1, 0xe760d676, 0xea23f0af, 0xeee2ed18,
	0xf0a5bd1d, 0xf464a0aa, 0xf9278673, 0xfde69bc4, 0x89b8fd09,
	0x8d79e0be, 0x803ac667, 0x84fbdbd0, 0x9abc8bd5, 0x9e7d9662,
	0x933eb0bb, 0x97ffad0c, 0xafb010b1, 0xab710d06, 0xa6322bdf,
	0xa2f33668, 0xbcb4666d, 0xb8757bda, 0xb5365d03, 0xb1f740b4
};
#define	COMPUTE(var, ch)	(var) = ((var) << 8) ^ crctab[((var) >> 24) ^ (ch)]

/*
 * Create a new mtree_cksum.
 */
struct mtree_cksum *
mtree_cksum_create(uint32_t init)
{
	struct mtree_cksum *cksum;

	cksum = calloc(1, sizeof(struct mtree_cksum));
	if (cksum != NULL)
		cksum->crc = init;
	return (cksum);
}

/*
 * Free the given mtree_cksum.
 */
void
mtree_cksum_free(struct mtree_cksum *cksum)
{

	assert(cksum != NULL);

	free(cksum);
}

/*
 * Reset the given mtree_cksum to its initial state.
 */
void
mtree_cksum_reset(struct mtree_cksum *cksum, uint32_t init)
{

	assert(cksum != NULL);

	cksum->crc  = init;
	cksum->len  = 0;
	cksum->done = 0;
}

/*
 * Update the checksum with the given data.
 */
void
mtree_cksum_update(struct mtree_cksum *cksum, const unsigned char *data, size_t len)
{

	assert(cksum != NULL);

	if (cksum->done)
		return;
	if (len == 0)
		return;

	assert(data != NULL);

	cksum->len += len;
	for (; len--; data++)
		COMPUTE(cksum->crc, *data);
}

/*
 * Finalize the checksum computation and retrieve the result.
 *
 * This function may be called multiple times to read the result, but no futher
 * calls to mtree_cksum_update() are possible.
 */
uint32_t
mtree_cksum_get_result(struct mtree_cksum *cksum)
{
	size_t len;

	assert(cksum != NULL);

	if (!cksum->done) {
		/*
		 * Include the length of the file and finalize.
		 */
		for (len = cksum->len; len != 0; len >>= 8)
			COMPUTE(cksum->crc, len & 0xFF);
		cksum->crc = ~cksum->crc;
		cksum->done = 1;
	}
	return (cksum->crc);
}

/*
 * Calculate checksum of bytes read from the given file descriptor and store
 * the result in crc.
 */
int
mtree_cksum_fd(int fd, uint32_t *crc)
{
	struct mtree_cksum	*cksum;
	unsigned char		 buf[16 * 1024];
	ssize_t			 n;

	assert(crc != NULL);

	cksum = mtree_cksum_create(MTREE_CKSUM_DEFAULT_INIT);
	if (cksum == NULL)
		return (-1);

	while ((n = read(fd, buf, sizeof(buf))) > 0)
		mtree_cksum_update(cksum, buf, n);
	if (n == -1) {
		mtree_cksum_free(cksum);
		return (-1);
	}
	*crc = mtree_cksum_get_result(cksum);

	mtree_cksum_free(cksum);
	return (0);
}

/*
 * Calculate checksum of bytes read from the given file path and store
 * the result in crc.
 */
int
mtree_cksum_path(const char *path, uint32_t *crc)
{
	int fd;
	int ret;

	assert(path != NULL);
	assert(crc != NULL);

	fd = open(path, O_RDONLY);
	if (fd != -1) {
		ret = mtree_cksum_fd(fd, crc);
		if (ret != 0) {
			int err = errno;

			close(fd);
			errno = err;
		} else
			close(fd);
		return (ret);
	}
	return (-1);
}
